nemo/collections/asr/parts/utils/decoder_timestamps_utils.py

```diff
@@ -0,0 +1,761 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import math
+import sys
+from datetime import datetime
+from typing import Dict, List, Tuple, Type, Union
+
+import librosa
+import numpy as np
+import soundfile as sf
+import torch
+from omegaconf import OmegaConf
+
+import nemo.collections.asr as nemo_asr
+from nemo.collections.asr.metrics.wer import WER
+from nemo.collections.asr.metrics.wer_bpe import WERBPE
+from nemo.collections.asr.models import EncDecCTCModel, EncDecCTCModelBPE
+from nemo.collections.asr.parts.utils.speaker_utils import audio_rttm_map, get_uniqname_from_filepath
+from nemo.collections.asr.parts.utils.streaming_utils import AudioFeatureIterator, FrameBatchASR
+from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
+from nemo.utils import logging
+
+__all__ = ['ASR_TIMESTAMPS']
+
+try:
+    from pyctcdecode import build_ctcdecoder
+except ImportError:
+    logging.info("pyctcdecode is not installed. You must install kenlm and pyctcdecode to use language models")
+
+
+def if_none_get_default(param, default_value):
+    return (param, default_value)[param is None]
+
+
+class WERBPE_TS(WERBPE):
+    """
+    This is WERBPE_TS class that is modified for generating word_timestamps with logits.
+    The functions in WER class is modified to save the word_timestamps whenever BPE token
+    is being saved into a list.
+    This class is designed to support ASR models based on CTC and BPE.
+    Please refer to the definition of WERBPE class for more information.
+    """
+
+    def __init__(
+        self,
+        tokenizer: TokenizerSpec,
+        batch_dim_index=0,
+        use_cer=False,
+        ctc_decode=True,
+        log_prediction=True,
+        dist_sync_on_step=False,
+    ):
+
+        super().__init__(tokenizer, batch_dim_index, use_cer, ctc_decode, log_prediction, dist_sync_on_step)
+
+    def ctc_decoder_predictions_tensor_with_ts(
+        self, time_stride, predictions: torch.Tensor, predictions_len: torch.Tensor = None
+    ) -> List[str]:
+        hypotheses, timestamps, word_timestamps = [], [], []
+        # '⁇' string should removed since it causes error on string split.
+        unk = '⁇'
```

Comment:
why the hypothesis contains '??' what it represents?

Reply:
'⁇' string is generated from decoder and this causes error during string process. This should be removed to make decoder function work smoothly.